File: Algorithms/Medium/DesignLinkedList.py
class Node:
    def __init__(self, val, next=None):
        self.val = val
        self.next = next


class MyLinkedList:
    def __init__(self):
        self.head = None
        self.tail = None
        self.len = 0

    # Time complexity: O(n)
    def get(self, index: int) -> int:
        if index > self.len - 1:
            return -1

        curr = self.head
        while index > 0:
            curr = curr.next
            index -= 1

        return curr.val

    # Time complexity: O(1)
    def addAtHead(self, val: int) -> None:
        self.len += 1
        self.head = Node(val, self.head)
        if self.tail is None:
            self.tail = self.head

    # Time complexity: O(1)
    def addAtTail(self, val: int) -> None:
        if self.head is None and self.tail is None:
            return self.addAtHead(val)

        self.len += 1
        new_node = Node(val)
        self.tail.next = new_node
        self.tail = new_node

    # Time complexity: O(n)
    def addAtIndex(self, index: int, val: int) -> None:
        if index > self.len:
            return

        if index == self.len:
            return self.addAtTail(val)

        if index == 0:
            return self.addAtHead(val)

        self.len += 1
        prev = Node(0, self.head)
        while index > 0:
            prev = prev.next
            index -= 1

        nxt = prev.next
        prev.next = Node(val, nxt)

    # Time complexity: O(n)
    def deleteAtIndex(self, index: int) -> None:
        if index > self.len - 1:
            return

        self.len -= 1
        if index == 0:
            self.head = self.head.next
            return

        prev = Node(0, self.head)
        while index > 0:
            prev = prev.next
            index -= 1

        if prev.next == self.tail:
            self.tail = prev

        prev.next = prev.next.next

File: Algorithms/Medium/707.Design-Linked-List/DesignLinkedList.go
package medium

type ListNode struct {
	Val  int
	Next *ListNode
}

func NewNode(val int, next *ListNode) *ListNode {
	return &ListNode{Val: val, Next: next}
}

type MyLinkedList struct {
	Head *ListNode
	Tail *ListNode
	Len  int
}

func NewLinkedList() MyLinkedList {
	return MyLinkedList{}
}

// Time complexity: O(n)
func (this *MyLinkedList) Get(index int) int {
	if index > this.Len-1 {
		return -1
	}

	curr := this.Head
	for index > 0 {
		curr = curr.Next
		index -= 1
	}

	return curr.Val
}

// Time complexity: O(1)
func (this *MyLinkedList) AddAtHead(val int) {
	this.Len += 1
	this.Head = NewNode(val, this.Head)
	if this.Tail == nil {
		this.Tail = this.Head
	}
}

// Time complexity: O(1)
func (this *MyLinkedList) AddAtTail(val int) {
	if this.Tail == nil && this.Head == nil {
		this.AddAtHead(val)
		return
	}

	this.Len += 1
	this.Tail.Next = NewNode(val, nil)
	this.Tail = this.Tail.Next
}

// Time complexity: O(n)
func (this *MyLinkedList) AddAtIndex(index int, val int) {
	if index > this.Len {
		return
	}

	if index == this.Len {
		this.AddAtTail(val)
		return
	}

	if index == 0 {
		this.AddAtHead(val)
		return
	}

	this.Len += 1
	curr := NewNode(0, this.Head)
	for index > 0 {
		curr = curr.Next
		index -= 1
	}

	nxt := curr.Next
	curr.Next = NewNode(val, nxt)
}

// Time complexity: O(n)
func (this *MyLinkedList) DeleteAtIndex(index int) {
	if index > this.Len-1 {
		return
	}

	this.Len -= 1
	if index == 0 {
		this.Head = this.Head.Next
		return
	}

	prev := NewNode(0, this.Head)
	for index > 0 {
		prev = prev.Next
		index -= 1
	}

	if prev.Next == this.Tail {
		this.Tail = prev
	}

	prev.Next = prev.Next.Next
}
